src/jit/lowerarmarch.cpp

```diff
@@ -693,7 +693,35 @@ void Lowering::ContainCheckCast(GenTreeCast* node)
 //
 void Lowering::ContainCheckCompare(GenTreeOp* cmp)
 {
-    ContainCheckBinary(cmp);
+    if (CheckImmedAndMakeContained(cmp, cmp->gtOp2))
+    {
+#ifdef _TARGET_ARM64_
+        GenTreePtr op1 = cmp->gtOp.gtOp1;
+        GenTreePtr op2 = cmp->gtOp.gtOp2;
+
+        // If op1 codegen can set flags op2 is an immediate 0
+        // we don't need to generate cmp instruction,
+        // provided we don't have another GenTree node between op1
+        // and cmp that could potentially modify flags.
+        //
+        // TODO-CQ: right now the below peep is inexpensive and
+        // gets the benefit in most of cases because in majority
+        // of cases op1, op2 and cmp would be in that order in
+        // execution.  In general we should be able to check that all
+        // the nodes that come after op1 in execution order do not
+        // modify the flags so that it is safe to avoid generating a
+        // test instruction.  Such a check requires that on each
+        // GenTree node we need to set the info whether its codegen
+        // will modify flags.
+        if (op2->IsIntegralConst(0) && (op1->gtNext == op2) && (op2->gtNext == cmp) &&
+            cmp->OperIs(GT_EQ, GT_NE, GT_GT, GT_GE, GT_LT, GT_LE) && op1->OperIs(GT_ADD, GT_AND, GT_SUB))
```

Comment:
Instead of `cmp->OperIs(GT_EQ, GT_NE, GT_GT, GT_GE, GT_LT, GT_LE)` we typically use `cmp->OperIsCompare()` , I note that it now also contains TEST_EQ and TEST_NE

Reply:
I'm not a big fan of large OperIs(...) calls:
```
    template <typename... T>
    bool OperIs(genTreeOps oper, T... rest)
    {
        return OperIs(oper) || OperIs(rest...);
    }
```

Reply:
OK.  I was hoping it was more efficient than that.

Reply:
You should check how the TEST_EQ and TEST_NE cases shodul be handled here.

Reply:
> I'm not a big fan of large OperIs(...) calls:

When I introduced `OperIs` I verified that C++ compilers handle it as expected. For example, this particular use ends up as `sub`/`cmp`/`ja` (which happens to be more efficient than `OperIsCompare` but that's another story).

Reply:
It would still be expanded into 6 consecutive compare|subtract and branch sequences right?

What was/is the issue with OperIsCompare ?
It should be one memory fetch and a bit test right?

    static bool OperIsCompare(genTreeOps gtOper)
    {
        return (OperKind(gtOper) & GTK_RELOP) != 0;
    }
    static unsigned OperKind(unsigned gtOper)
    {
        assert(gtOper < GT_COUNT);

        return gtOperKindTable[gtOper];
    }

Reply:
> It would still be expanded into 6 consecutive compare|subtract and branch sequences right?

Nope. GT_EQ & co. form a continuous range so it's something like `if ((x - GT_EQ) < 6)`.

> What was/is the issue with OperIsCompare ?

Yes, one memory fetch. Even if the memory is in cache it's still 4-5 cycles, that's already more than the 2 cycles (sub+cmp) for the `OperIs` version. And the bit test adds another cycle.